### ISSUE TITLE
Long delay with a large number of objects and rules for them

### ISSUE THREAD

**Issue by vitorli** [algenty/grafana-flowcharting#289]
Good day!

At the moment, I began to observe a long delay in the flowcharting response with a large number of objects and rules for them, can I somehow get around this? Maybe you should add resources for better performance? Because I need large circuits and their processing.

If there is a solution or advice, I would be grateful!

**Comment by CrashLaker** [comment]
hi @vitorli 

can you tell if your delay is due to metric queries download to complete?
i've also got some pretty dense diagrams and they won't show up unless all metrics are downloaded.

**Comment by vitorli** [comment]
Hi @CrashLaker
I don't have a lot of metrics that are used, basically, a large number of objects and rules related to them (1500 objects and about 350 rules). Even simple editing takes a long time to load everything. Moreover, the hangs start after 100 rules stably.

**Comment by vitorli** [comment]
![image](https://user-images.githubusercontent.com/48572627/116215442-5cbc5200-a750-11eb-83b6-4632757b5837.png)

**Comment by xkilian** [comment]
Yeah, it is not metric related, nor the complexity of the XML drawing. 

It is specifically related to the number of rules. I have not seen any improvements with disabling things likes regexes, having no links. Just associating metrics, thresholds and colours. Around 80 rules, it will gradually and noticeably slow down as you add more and more rules.
Could it be related to rendering the rules sections for display in Grafana?
Could it be related to how the rules are trying to dynamically modify the XML?
I dunno.
@algenty would need to do some performance tracing to see what is actually slowing it down, when you start adding more rules.

**Comment by vitorli** [comment]
In my project there are 350 rules, in general, everything takes a very long time to load, and if you need to make adjustments to the rules, it’s just unbearably long, sometimes the page just freezes and resets the changes when reloading and you need to start over
![image](https://user-images.githubusercontent.com/48572627/116247880-f6e1c180-a773-11eb-8ce1-422703b06d49.png)

**Comment by xkilian** [comment]
You have to disable automatic reloads..

**Comment by vitorli** [comment]
Перезагрузки отключены, я имел ввиду, что страница не отвечает и зависает

**Comment by vitorli** [comment]
Reloads are disabled, I meant that the page does not respond and freezes

**Comment by meshkov** [comment]
I can cofirm this problem on large number rules.
Dashboard loads 20-30sec.
![image](https://user-images.githubusercontent.com/1703670/116398349-5dc4b080-a830-11eb-9f32-6b489403962a.png)

**Comment by algenty** [comment]
Thx @xkilian and @xkilian for you support during this pause.
Next version will be better for response time.
Since few month, i have upgraded with rxjs event but there are too many bugs at this time to publish it.

**Comment by Thuaiqian** [comment]
Mine is the latest version, and I still have the problem of loading slowly due to too many rules. Is there a workaround？

**Comment by pauljones0** [comment]
While this isn't the same issue, I have noticed that when editing the flow charting plugin, the memory allocation keeps growing and growing and I've had pages crash because of this.
